File: 백준/Silver/1158. 요세푸스 문제/요세푸스 문제.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        List<Integer> list = new LinkedList<>();
        StringBuilder sb = new StringBuilder();
        sb.append("<");
        for (int i = 1; i <= n; i++) list.add(i);
        int cur = m - 1;
        while (!list.isEmpty()) {
            cur %= list.size();
            int tmp = list.remove(cur);
            sb.append(tmp);
            if (!list.isEmpty()) sb.append(", ");
            cur += m - 1;
        }
        sb.append(">");
        System.out.print(sb);
    }
}

File: 백준/Silver/1158. 요세푸스 문제/요세푸스 문제.kt
import java.util.*

/**
 * 풀이: 두 가지
 * 1. 큐를 이용하여 원형 순회 구현하기
 * 2. 연결 리스트 흉내내기
 */

val left = IntArray(5001) { i -> i - 1 }
val right = IntArray(5001) { i -> i + 1 }

fun remove(idx: Int) {
    val originRight = right[idx]
    val originLeft = left[idx]
    right[originLeft] = originRight
    left[originRight] = originLeft
}

fun move(start: Int, cnt: Int): Int {
    var nxt = start
    repeat(cnt) {
        nxt = right[nxt]
    }

    return nxt
}

fun main() = with(System.`in`.bufferedReader()) {
    val (n, k) = readLine().split(" ").map { it.toInt() }
    left[1] = n
    right[n] = 1

    val list = mutableListOf<Int>()
    var target = 0
    repeat(n) {
        target = move(target, k)
        list.add(target)
        remove(target)
    }

    print(list.joinToString(", ", "<", ">"))
}